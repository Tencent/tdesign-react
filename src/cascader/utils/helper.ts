import { TreeNode, CascaderContextType, TreeNodeValue, TreeOptionData } from '../interface';

/**
 * 获取node的全部路径
 * @param node
 * @returns
 */
export function getFullPathLabel(node: TreeNode) {
  return node
    .getPath()
    .map((node: TreeNode) => node.label)
    .join('/');
}

/**
 * treeValue计算方法
 * @param value
 * @returns
 */
export const getTreeValue = (value: CascaderContextType['value']) => {
  let treeValue: TreeNodeValue[] = [];
  if (Array.isArray(value)) {
    if (value.length > 0 && typeof value[0] === 'object') {
      treeValue = (value as TreeOptionData[]).map((val) => val.value);
    } else if (value.length) {
      treeValue = value as TreeNodeValue[];
    }
  } else if (value) {
    if (typeof value === 'object') {
      treeValue = [(value as TreeOptionData).value];
    } else {
      treeValue = [value];
    }
  }
  return treeValue;
};

/**
 * 判断node是否在页面上
 * @param node Node
 */
export function isInPage(node: Node) {
  return node === document.body ? false : document.body.contains(node);
}

export default {
  getFullPathLabel,
};
