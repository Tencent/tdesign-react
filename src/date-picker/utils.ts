import chunk from 'lodash/chunk';
import { DateValue } from './type';
import { Styles } from '../common';

/**
 * next Month
 * @since version V1.0.0
 * @param { Date } date
 * @returns {Date}
 */
export const nextMonth = (date: Date) => {
  const isPassYear = date.getMonth() === 11;
  const newMonth = isPassYear ? 0 : date.getMonth() + 1;
  const year = isPassYear ? date.getFullYear() + 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * prev Month
 * @since version V1.0.0
 * @param {Date} date
 * @returns {Date}
 */
export const prevMonth = (date: Date) => {
  const passYear = date.getMonth() === 0;
  const newMonth = passYear ? 11 : date.getMonth() - 1;
  const year = passYear ? date.getFullYear() - 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * dateIndexOf
 * @since version V1.0.0
 * @param {Date} date
 * @param {Date[]} arrDates
 * @returns {Number}
 */
export function dateIndexOf(date: Date, arrDates: DateValue[]): number {
  const curDate = new Date(date);
  const curTime = curDate.setHours(0, 0, 0, 0);
  const selectedDates = arrDates.map((d: Date) => {
    const dt = new Date(d);
    const dtime = dt.setHours(0, 0, 0, 0);
    return dtime;
  });

  // console.log('curDate', selectedDates, curDate, selectedDates.indexOf(curTime));

  return selectedDates.indexOf(curTime);
}

/**
 * Empty value
 * @since version V1.0.0
 * @param mixedVar
 * @returns {Boolean}
 * @example
 * // example 1: empty(null)
 * // returns 1: true
 * // example 2: empty(undefined)
 * // returns 2: true
 * // example 3: empty([])
 * // returns 3: true
 * // example 4: empty({})
 * // returns 4: true
 * // example 5: empty({'fn' : function () {} })
 * // returns 5: false
 */
export function empty(mixedVar: any): mixedVar is undefined | null | false | '' | {} {
  let undef: undefined;
  const emptyValues = [undef, null, false, ''];
  for (let i = 0, len = emptyValues.length; i < len; i++) {
    if (mixedVar === emptyValues[i]) {
      return true;
    }
  }

  if (typeof mixedVar === 'object') {
    return Object.keys(mixedVar).length < 1;
  }

  return false;
}

/**
 * To Dashed from Camel Case
 * @since version V1.0.0
 * @param {String} strCamelCase
 * @returns {string}
 * @example
 * toDash('strCamelCase') === 'str-camel-case'
 */
export function toDash(strCamelCase: string): string {
  return String(strCamelCase)
    .split(/(?=[A-Z])/)
    .join('-')
    .toLowerCase();
}

export function setStyles(el: HTMLElement | SVGElement, styles: Styles) {
  Object.entries(styles).forEach(([prop, val = null]) => {
    if (empty(val)) {
      el.style.removeProperty(prop);
      return;
    }
    const [value, pri = ''] = String(val).split('!');

    el.style.setProperty(toDash(prop), value, pri);
  });
}

/**
 * 首字母大写
 * @param {String} str 目标字符串
 * @returns {String}
 */
export function firstUpperCase(str: string): string {
  if (!str) return str;
  return str[0].toUpperCase().concat(str.substring(1, str.length));
}

interface DateObj {
  year: number;
  month: number;
}

/**
 * 返回指定年、月的第一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getFirstDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, 1);
}

/**
 * 返回指定年、月的天数
 * @param {Object} { year, month }
 * @returns {Number}
 */
function getDaysInMonth({ year, month }: DateObj): number {
  return new Date(year, month + 1, 0).getDate();
}

/**
 * 返回指定年、月的最后一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getLastDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, getDaysInMonth({ year, month }));
}

function isSameYear(date1: Date, date2: Date) {
  return date1.getFullYear() === date2.getFullYear();
}

function isSameMonth(date1: Date, date2: Date) {
  return isSameYear(date1, date2) && date1.getMonth() === date2.getMonth();
}

function isSameDate(date1: Date, date2: Date) {
  return isSameMonth(date1, date2) && date1.getDate() === date2.getDate();
}

/**
 * 是否是某法范围内的日期，精确到日
 * @param {Date} value 目标日期
 * @param {Object} { start, end } 范围
 * @returns {Boolean}
 */
function isBetween(
  value: { getFullYear: () => number; getMonth: () => number; getDate: () => number },
  { start, end }: { start: any; end: any },
): boolean {
  const date = new Date(value.getFullYear(), value.getMonth(), value.getDate());

  const startTime = new Date(start.getFullYear(), start.getMonth(), start.getDate());
  const endTime = new Date(end.getFullYear(), end.getMonth(), end.getDate());

  return startTime <= date && endTime >= date;
}

/**
 * 比较两个日期对象的时间戳
 * @param {Date} date1 日期1
 * @param {Date} date2 日期2
 * @returns {Number} 返回 date1.getTime() - date2.getTime() 的差值
 */
function compareAsc(date1: { getTime: () => any }, date2: Date): number {
  const d1 = date1.getTime();
  const d2 = date2.getTime();

  if (d1 < d2) return -1;
  if (d1 > d2) return 1;
  return 0;
}

/**
 * 比较两个 Date 是否是同一天 或则 同一月 或则 同一年
 * @param {Date} date1 比较的日期
 * @param {Date} date2 比较的日期
 * @param {String} type 比较类型，默认比较到『日』 date|month|year
 * @returns {Boolean}
 */
export function isSame(date1: Date, date2: Date, type = 'date'): boolean {
  const func = {
    isSameYear,
    isSameMonth,
    isSameDate,
  };
  return func[`isSame${firstUpperCase(type)}`](date1, date2);
}

export function outOfRanges(d: Date, min: any, max: any) {
  return (min && compareAsc(d, min) === -1) || (max && compareAsc(d, max) === 1);
}

/**
 * @returns {Date} 当天零点的日期对象
 */
export function getToday(): Date {
  const now = new Date();
  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0);
}

/**
 * 返回日期对象的年、月、日、小时、分钟、秒、12小时制标识
 * @param {Date} date
 * @returns {Object}
 */
export function getDateObj(date: Date) {
  if (!(date instanceof Date)) {
    throw new Error('date must be Date object');
  }
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    date: date.getDate(),
    hours: date.getHours(),
    minutes: date.getMinutes(),
    seconds: date.getSeconds(),
    meridiem: date.getHours() > 11 ? 'PM' : 'AM',
  };
}

/**
 * 设置日期对象的时间部分
 * @param {Date} d 日期
 * @param {Number} hour 小时
 * @param {Number} min 分钟
 * @param {Number} sec 秒
 * @returns {Date} 一个新的date
 */
export function setDateTime(d: Date, hour: number, min: number, sec: number): Date {
  // eslint-disable-next-line
  const { year, month, date } = getDateObj(d);
  return new Date(year, month, date, hour, min, sec, 0);
}

/**
 * 增加月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function subtractMonth(date: Date, num: any): Date {
  const day = date.getDate();
  const newDate = new Date(date);

  let monthNum = num;
  // eslint-disable-next-line no-plusplus
  while (monthNum--) {
    newDate.setDate(0);
  }
  newDate.setDate(day);
  return newDate;
}

/**
 * 减月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function addMonth(date: Date, num: number): Date {
  let monthNum = num;
  if (num < 0) monthNum = 0;
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() + monthNum);
  return newDate;
}

export interface OptionsType {
  firstDayOfWeek: number;
  disableDate: Function;
  minDate: Date;
  maxDate: Date;
  monthText?: string;
}

export function getWeeks(
  { year, month }: { year: number; month: number },
  { firstDayOfWeek, disableDate = () => false, minDate, maxDate }: OptionsType,
) {
  const prependDay = getFirstDayOfMonth({ year, month });

  const appendDay = getLastDayOfMonth({ year, month });

  const maxDays = getDaysInMonth({ year, month });

  const daysArr = [];

  const today = getToday();

  for (let i = 1; i <= maxDays; i++) {
    const currentDay = new Date(year, month, i);

    daysArr.push({
      text: i,
      active: false,
      value: currentDay,
      disabled: disableDate(currentDay) || outOfRanges(currentDay, minDate, maxDate),
      now: isSame(today, currentDay),
      firstDayOfMonth: i === 1,
      lastDayOfMonth: i === maxDays,
      type: 'current-month',
    });
  }

  if (prependDay.getDay() !== firstDayOfWeek - 1) {
    prependDay.setDate(0); // 上一月

    const FLAG = true;
    while (FLAG) {
      daysArr.unshift({
        text: prependDay.getDate().toString(),
        active: false,
        value: new Date(prependDay),
        disabled: disableDate(prependDay) || outOfRanges(prependDay, minDate, maxDate),
        additional: true, // 非当前月
        type: 'prev-month',
      });
      prependDay.setDate(prependDay.getDate() - 1);
      if (prependDay.getDay() === Math.abs(firstDayOfWeek + 6) % 7) break;
    }
  }

  const LEN = 42; // 显示6周

  while (daysArr.length < LEN) {
    appendDay.setDate(appendDay.getDate() + 1);
    daysArr.push({
      text: appendDay.getDate(),
      active: false,
      value: new Date(appendDay),
      disabled: disableDate(appendDay) || outOfRanges(appendDay, minDate, maxDate),
      additional: true, // 非当前月
      type: 'next-month',
    });
  }

  return chunk(daysArr, 7);
}

export function getYears(year: number, { disableDate = () => false, minDate, maxDate }: OptionsType) {
  const startYear = parseInt((year / 10).toString(), 10) * 10;
  const endYear = startYear + 9;

  const yearArr = [];

  const today = getToday();

  for (let i = startYear; i <= endYear; i++) {
    const date = new Date(i, 1);

    let disabledMonth = 0;
    let outOfRangeMonth = 0;

    for (let j = 0; j < 12; j++) {
      const d = new Date(i, j);
      if (disableDate(d)) disabledMonth += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeMonth += 1;
    }

    yearArr.push({
      value: date,
      now: isSame(date, today, 'year'),
      disabled: disabledMonth === 12 || outOfRangeMonth === 12,
      active: false,
      text: `${date.getFullYear()}`,
    });
  }

  return chunk(yearArr, 4);
}

export function getMonths(year: number, { disableDate = () => false, minDate, maxDate, monthText = '' }: OptionsType) {
  const monthArr = [];

  const today = getToday();

  for (let i = 0; i <= 11; i++) {
    const date = new Date(year, i);

    let disabledDay = 0;
    let outOfRangeDay = 0;

    const daysInMonth = getDaysInMonth({ year, month: i });

    for (let j = 1; j <= daysInMonth; j++) {
      const d = new Date(year, i, j);
      if (disableDate(d)) disabledDay += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeDay += 1;
    }

    monthArr.push({
      value: date,
      now: isSame(date, today, 'month'),
      disabled: disabledDay === daysInMonth || outOfRangeDay === daysInMonth,
      active: false,
      text: `${date.getMonth() + 1} ${monthText}`,
    });
  }

  return chunk(monthArr, 4);
}

interface DateTime {
  active: boolean;
  highlight: boolean;
  startOfRange: boolean;
  endOfRange: boolean;
  value: Date;
}

export function flagActive(data: any[], { ...args }: any) {
  const { start, end, type = 'date' } = args;

  if (!end) {
    return data.map((row: any[]) =>
      row.map((item: DateTime) => {
        Object.assign(item, { active: isSame(item.value, start, type) });
        return item;
      }),
    );
  }

  return data.map((row: any[]) =>
    row.map((item: DateTime) => {
      const date = item.value;
      const isStart = isSame(start, date, type);
      const isEnd = isSame(end, date, type);
      Object.assign(item, {
        active: isStart || isEnd,
        highlight: isBetween(date, { start, end }),
        startOfRange: isStart,
        endOfRange: isEnd,
      });
      return item;
    }),
  );
}
