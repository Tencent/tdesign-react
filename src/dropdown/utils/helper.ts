import React from 'react';
import { DropdownOption } from '../type';
import DropdownMenu from '../DropdownMenu';

export const getOptionsFromChildren = (children: React.ReactElement): DropdownOption[] => {
  if (children.type === DropdownMenu) {
    const groupChildren = children.props.children as React.ReactElement;
    if (Array.isArray(groupChildren)) {
      return getOptionsFromChildren(groupChildren);
    }
  }

  if (Array.isArray(children)) {
    return children.map((item: React.ReactElement) => {
      const groupChildren = item.props.children;
      if (Array.isArray(groupChildren)) {
        const contentIdx = groupChildren.findIndex((v) => typeof v === 'string');
        return {
          ...item.props,
          content: groupChildren[contentIdx],
          children: getOptionsFromChildren(groupChildren[1]),
        };
      }
      return { ...item.props, content: item.props.children || item.props.content };
    });
  }

  return [];
};
