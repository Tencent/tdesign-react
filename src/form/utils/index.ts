import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
