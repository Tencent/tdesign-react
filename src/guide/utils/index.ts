import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
