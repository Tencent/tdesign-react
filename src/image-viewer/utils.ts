export const downloadFile = function (imgSrc) {
  const image = new Image();
  const name = imgSrc?.split?.('/').pop() || Math.random().toString(32).slice(2);

  image.setAttribute('crossOrigin', 'anonymous');

  image.onload = () => {
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;

    const context = canvas.getContext('2d');
    context.drawImage(image, 0, 0, image.width, image.height);
    canvas.toBlob((blob) => {
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.download = name;
      a.href = url;
      a.click();
      a.remove();
      URL.revokeObjectURL(url);
    });
  };
  image.src = imgSrc;
};
