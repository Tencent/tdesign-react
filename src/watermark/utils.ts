const toLowercaseSeparator = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

// style对象转字符串
export const getStyleStr = (style: React.CSSProperties): string =>
  Object.keys(style)
    .map((key: keyof React.CSSProperties) => `${toLowercaseSeparator(key)}: ${style[key]};`)
    .join(' ');
