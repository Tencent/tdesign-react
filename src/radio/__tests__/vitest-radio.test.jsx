/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * 该文件由脚本自动生成，如需修改请联系 PMC
 * This file generated by scripts of tdesign-api. `npm run api:docs Radio React(PC) vitest,finalProject`
 * If you need to modify this file, contact PMC first please.
 */
import React from 'react';
import { fireEvent, vi, render } from '@test/utils';
import { Radio, RadioGroup } from '..';
import { getRadioGroupKidsMount, getRadioGroupDefaultMount } from './mount';

describe('Radio Component', () => {
  it('props.allowUncheck works fine', () => {
    const onChangeFn = vi.fn();
    const { container } = render(<Radio checked={true} allowUncheck={true} onChange={onChangeFn}></Radio>);
    fireEvent.click(container.firstChild);
    expect(onChangeFn).toHaveBeenCalled();
    expect(onChangeFn.mock.calls[0][0]).toBe(false);
    expect(onChangeFn.mock.calls[0][1].e.type).toBe('click');
  });

  it('props.checked works fine', () => {
    // checked default value is false
    const { container: container1 } = render(<Radio></Radio>);
    expect(container1.querySelector(`.${'t-is-checked'}`)).toBeFalsy();
    // checked = true
    const { container: container2 } = render(<Radio checked={true}></Radio>);
    expect(container2.firstChild).toHaveClass('t-is-checked');
    expect(container2).toMatchSnapshot();
    // checked = false
    const { container: container3 } = render(<Radio checked={false}></Radio>);
    expect(container3.querySelector(`.${'t-is-checked'}`)).toBeFalsy();
    expect(container3).toMatchSnapshot();
  });

  it(`props.checked is equal to true`, () => {
    const { container } = render(<Radio checked={true}></Radio>);
    const domWrapper = container.querySelector('input');
    expect(domWrapper.checked).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('props.children works fine', () => {
    const { container } = render(
      <Radio>
        <span className="custom-node">TNode</span>
      </Radio>,
    );
    expect(container.querySelector('.custom-node')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('props.disabled works fine', () => {
    // disabled default value is undefined
    const { container: container1 } = render(<Radio>Text</Radio>);
    expect(container1.querySelector(`.${'t-is-disabled'}`)).toBeFalsy();
    // disabled = true
    const { container: container2 } = render(<Radio disabled={true}>Text</Radio>);
    expect(container2.firstChild).toHaveClass('t-is-disabled');
    expect(container2).toMatchSnapshot();
    // disabled = false
    const { container: container3 } = render(<Radio disabled={false}>Text</Radio>);
    expect(container3.querySelector(`.${'t-is-disabled'}`)).toBeFalsy();
    expect(container3).toMatchSnapshot();
  });

  it('props.disabled works fine', () => {
    const onChangeFn = vi.fn();
    const { container } = render(
      <Radio disabled={true} onChange={onChangeFn}>
        Text
      </Radio>,
    );
    fireEvent.click(container.firstChild);
    expect(onChangeFn).not.toHaveBeenCalled();
  });

  it('props.label works fine', () => {
    const { container } = render(<Radio label={<span className="custom-node">TNode</span>}></Radio>);
    expect(container.querySelector('.custom-node')).toBeTruthy();
    expect(container.querySelector('.t-radio__label')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('props.name works fine', () => {
    const wrapper = render(<Radio name="radio-gender-name"></Radio>);
    const container = wrapper.container.querySelector('input');
    expect(container.getAttribute('name')).toBe('radio-gender-name');
  });

  it(`props.value is equal to 'radio-value'`, () => {
    const { container } = render(<Radio value="radio-value"></Radio>);
    const domWrapper = container.querySelector('input');
    expect(domWrapper.value).toBe('radio-value');
  });

  it('events.click works fine', () => {
    const onClickFn = vi.fn();
    const { container } = render(<Radio onClick={onClickFn}></Radio>);
    fireEvent.click(container.firstChild);
    expect(onClickFn).toHaveBeenCalled();
    expect(onClickFn.mock.calls[0][0].e.stopPropagation).toBeTruthy();
  });
});

describe('RadioGroup Component', () => {
  it('props.allowUncheck works fine', () => {
    const onChangeFn = vi.fn();
    const { container } = getRadioGroupDefaultMount(
      RadioGroup,
      { value: 1, allowUncheck: true },
      { onChange: onChangeFn },
    );
    fireEvent.click(container.querySelector('.t-radio'));
    expect(onChangeFn).toHaveBeenCalled();
    expect(onChangeFn.mock.calls[0][0]).toBe(undefined);
    expect(onChangeFn.mock.calls[0][1].e.type).toBe('click');
  });

  it('props.allowUncheck works fine', () => {
    const onChangeFn = vi.fn();
    const { container } = getRadioGroupKidsMount(
      RadioGroup,
      { value: 1, allowUncheck: true },
      { onChange: onChangeFn },
    );
    fireEvent.click(container.querySelector('.t-radio'));
    expect(onChangeFn).toHaveBeenCalled();
    expect(onChangeFn.mock.calls[0][0]).toBe(undefined);
    expect(onChangeFn.mock.calls[0][1].e.type).toBe('click');
  });

  it('props.disabled is equal true', () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup, { disabled: true });
    expect(container.querySelectorAll('.t-radio.t-is-disabled').length).toBe(4);
    expect(container).toMatchSnapshot();
  });

  it('props.disabled is equal true', () => {
    const { container } = getRadioGroupKidsMount(RadioGroup, { disabled: true });
    expect(container.querySelectorAll('.t-radio.t-is-disabled').length).toBe(4);
    expect(container).toMatchSnapshot();
  });

  it('props.disabled: disabled radio can not trigger change', () => {
    const onChangeFn = vi.fn();
    const { container } = getRadioGroupDefaultMount(RadioGroup, { disabled: true }, { onChange: onChangeFn });
    fireEvent.click(container.querySelector('.t-radio'));
    expect(onChangeFn).not.toHaveBeenCalled();
  });

  it('props.disabled: disabled radio can not trigger change', () => {
    const onChangeFn = vi.fn();
    const { container } = getRadioGroupKidsMount(RadioGroup, { disabled: true }, { onChange: onChangeFn });
    fireEvent.click(container.querySelector('.t-radio'));
    expect(onChangeFn).not.toHaveBeenCalled();
  });

  it(`props.name is equal to 'custom-radio-name'`, () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup, { name: 'custom-radio-name' });
    const domWrapper = container.querySelector('input');
    expect(domWrapper.getAttribute('name')).toBe('custom-radio-name');
    expect(container).toMatchSnapshot();
  });

  it(`props.name is equal to 'custom-radio-name'`, () => {
    const { container } = getRadioGroupKidsMount(RadioGroup, { name: 'custom-radio-name' });
    const domWrapper = container.querySelector('input');
    expect(domWrapper.getAttribute('name')).toBe('custom-radio-name');
    expect(container).toMatchSnapshot();
  });

  it('props.options works fine. `{".t-radio":4}` should exist', () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup);
    expect(container.querySelectorAll('.t-radio').length).toBe(4);
  });

  it('props.options works fine. `".custom-node"` should exist', () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup);
    expect(container.querySelector('.custom-node')).toBeTruthy();
  });

  it('props.options works fine. `{".t-radio.t-is-disabled":1}` should exist', () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup);
    expect(container.querySelectorAll('.t-radio.t-is-disabled').length).toBe(1);
  });

  it(`props.value is equal to '2'`, () => {
    const { container } = getRadioGroupDefaultMount(RadioGroup, { value: '2' });
    const domWrapper = container.querySelector('.t-radio.t-is-checked input');
    expect(domWrapper.value).toBe('2');
  });

  it(`props.value is equal to '2'`, () => {
    const { container } = getRadioGroupKidsMount(RadioGroup, { value: '2' });
    const domWrapper = container.querySelector('.t-radio.t-is-checked input');
    expect(domWrapper.value).toBe('2');
  });

  const variantClassNameList = ['t-radio-group__outline', 't-radio-group--primary-filled', 't-radio-group--filled'];
  ['outline', 'primary-filled', 'default-filled'].forEach((item, index) => {
    it(`props.variant is equal to ${item}`, () => {
      const { container } = render(<RadioGroup variant={item}></RadioGroup>);
      expect(container.firstChild).toHaveClass(variantClassNameList[index]);
    });
  });
});
