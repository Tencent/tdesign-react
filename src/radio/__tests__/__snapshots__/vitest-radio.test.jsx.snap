// Vitest Snapshot v1

exports[`Radio Component > props.checked is equal to true 1`] = `
<div>
  <label
    class="t-radio t-is-checked"
  >
    <input
      checked=""
      class="t-radio__former"
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
  </label>
</div>
`;

exports[`Radio Component > props.checked works fine 1`] = `
<div>
  <label
    class="t-radio t-is-checked"
  >
    <input
      checked=""
      class="t-radio__former"
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
  </label>
</div>
`;

exports[`Radio Component > props.checked works fine 2`] = `
<div>
  <label
    class="t-radio"
  >
    <input
      class="t-radio__former"
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
  </label>
</div>
`;

exports[`Radio Component > props.children works fine 1`] = `
<div>
  <label
    class="t-radio"
  >
    <input
      class="t-radio__former"
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
    <span
      class="t-radio__label"
    >
      <span
        class="custom-node"
      >
        TNode
      </span>
    </span>
  </label>
</div>
`;

exports[`Radio Component > props.disabled works fine 1`] = `
<div>
  <label
    class="t-radio t-is-disabled"
  >
    <input
      class="t-radio__former"
      disabled=""
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
    <span
      class="t-radio__label"
    >
      Text
    </span>
  </label>
</div>
`;

exports[`Radio Component > props.disabled works fine 2`] = `
<div>
  <label
    class="t-radio"
  >
    <input
      class="t-radio__former"
      type="radio"
      value=""
    />
    <span
      class="t-radio__input"
    />
    <span
      class="t-radio__label"
    >
      Text
    </span>
  </label>
</div>
`;

exports[`RadioGroup Component > props.disabled is equal true 1`] = `
<div>
  <div
    class="t-radio-group t-size-m t-radio-group__outline"
  >
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="1"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio1
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="2"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio2
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="3"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        <span
          class="custom-node"
        >
          Radio3
        </span>
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="4"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio4
      </span>
    </label>
  </div>
</div>
`;

exports[`RadioGroup Component > props.disabled is equal true 2`] = `
<div>
  <div
    class="t-radio-group t-size-m t-radio-group__outline"
  >
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="1"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio1
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="2"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio2
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="3"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio3
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        type="radio"
        value="4"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio4
      </span>
    </label>
  </div>
</div>
`;

exports[`RadioGroup Component > props.name is equal to 'custom-radio-name' 1`] = `
<div>
  <div
    class="t-radio-group t-size-m t-radio-group__outline"
  >
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="1"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio1
      </span>
    </label>
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="2"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio2
      </span>
    </label>
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="3"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        <span
          class="custom-node"
        >
          Radio3
        </span>
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        name="custom-radio-name"
        type="radio"
        value="4"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio4
      </span>
    </label>
  </div>
</div>
`;

exports[`RadioGroup Component > props.name is equal to 'custom-radio-name' 2`] = `
<div>
  <div
    class="t-radio-group t-size-m t-radio-group__outline"
  >
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="1"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio1
      </span>
    </label>
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="2"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio2
      </span>
    </label>
    <label
      class="t-radio"
    >
      <input
        class="t-radio__former"
        name="custom-radio-name"
        type="radio"
        value="3"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio3
      </span>
    </label>
    <label
      class="t-radio t-is-disabled"
    >
      <input
        class="t-radio__former"
        disabled=""
        name="custom-radio-name"
        type="radio"
        value="4"
      />
      <span
        class="t-radio__input"
      />
      <span
        class="t-radio__label"
      >
        Radio4
      </span>
    </label>
  </div>
</div>
`;
